### PR TITLE
Add Snowflake DQ Operators 

### LINKED ISSUE #17694: Add SQL Check Operators to Snowflake Provider
Add SnowflakeCheckOperator, SnowflakeValueCheckOperator, and SnowflakeIntervalCheckOperator to Snowflake provider.

**Use case / motivation**

The SQL operator, as well as other provider DB operators, already support this functionality. It can prove useful for data quality use cases with Snowflake. It should be relatively easy to implement in the same fashion as [BigQuery's versions](https://github.com/apache/airflow/blob/main/airflow/providers/google/cloud/operators/bigquery.py#L101).

**Are you willing to submit a PR?**

Yep!

**Related Issues**

Not that I could find.

### PR DESCRIPTION
**Pull request by denimalpaca**
Add three new Snowflake operators based on SQL Checks

The SnowflakeCheckOperator, SnowflakeValueCheckOperator, and
SnowflakeIntervalCheckOperators are added as subclasses of their respective
SQL Operators. These additions follow the conventions set in the BigQueryOperators
subclassing from the same SQL_CheckOperators.

closes: #17694

### CODE REVIEW

**Comment by uranusjr** [comment]
I don’t understand Snowflake a bit to comment on whether this actually works, but the implementation seems straightforward (using things already present in Airflow) and sound.

**Comment by denimalpaca** [comment]
> I don’t understand Snowflake a bit to comment on whether this actually works, but the implementation seems straightforward (using things already present in Airflow) and sound.

If it makes you feel better, I did add [this DAG](https://github.com/astronomer/airflow-data-quality-demo/pull/2) that I'm working on concurrently to the breeze environment and saw it working in a real Snowflake environment. So two of the three operators did run successfully.

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
Reopened to rebuid!

**Comment by potiuk** [comment]
I think you need to rebase to latest main @denimalpaca

**Comment by potiuk** [comment]
Can you please rebase to the latest `main` ? I think there were some fixes to address genera CI job failures.

File: airflow/providers/snowflake/operators/snowflake.py [hunk 1]
```diff
@@ -125,3 +126,299 @@ def execute(self, context: Any) -> None:
...
+    depending on where you put it in your DAG, you have the choice to
+    stop the critical path, preventing from
+    publishing dubious data, or on the side and receive email alerts
+    without stopping the progress of the DAG.
+
+    :param sql: the sql code to be executed. (templated)
+    :type sql: Can receive a str representing a sql statement,
+        a list of str (sql statements), or reference to a template file.
+        Template reference are recognized by str ending in '.sql'
+    :param use_legacy_sql: Whether to use legacy SQL (true)
```
Comment (mik-laj) [inline airflow/providers/snowflake/operators/snowflake.py]: Is it used?

File: airflow/providers/snowflake/operators/snowflake.py [hunk 2]
```diff
@@ -125,3 +126,299 @@ def execute(self, context: Any) -> None:
...
+        self.query_ids = []
+
+
+class SnowflakeValueCheckOperator(_SnowflakeDbHookMixin, SQLValueCheckOperator):
+    """
+    Performs a simple value check using sql code.
+
+    :param sql: the sql to be executed
+    :type sql: str
+    :param use_legacy_sql: Whether to use legacy SQL (true)
```
Comment (mik-laj) [inline airflow/providers/snowflake/operators/snowflake.py]: Snowflake have only one SQL variant. BigQuery support legacy SQL and ZetaSQL.

Comment (denimalpaca) [reply airflow/providers/snowflake/operators/snowflake.py]: Ah, no. Mistakenly added as I used the BigQuery operator as a model.

File: airflow/providers/snowflake/operators/snowflake.py [hunk 3]
```diff
@@ -125,3 +126,284 @@ def execute(self, context: Any) -> None:
 
         if self.do_xcom_push:
             return execution_info
+
+
+class _SnowflakeDbHookMixin:
+    def get_db_hook(self) -> SnowflakeHook:
+        """
+        Create and return SnowflakeHook.
+        :return: a SnowflakeHook instance.
```
Comment (mik-laj) [inline airflow/providers/snowflake/operators/snowflake.py]:
```suggestion

        :return: a SnowflakeHook instance.
```
this is needed to render the documentation properly.

**Review by mik-laj — APPROVED** [review]

File: airflow/providers/snowflake/operators/snowflake.py [hunk 4]
```diff
@@ -125,3 +126,285 @@ def execute(self, context: Any) -> None:
 
         if self.do_xcom_push:
             return execution_info
+
+
+class _SnowflakeDbHookMixin:
```
Comment (JavierLopezT) [inline airflow/providers/snowflake/operators/snowflake.py]: Why a class and not just a method inside the SnowflakeCheckOperator?

File: airflow/providers/snowflake/operators/snowflake.py [hunk 5]
```diff
@@ -125,3 +126,285 @@ def execute(self, context: Any) -> None:
...
+    without stopping the progress of the DAG.
+
+    :param sql: the sql code to be executed. (templated)
+    :type sql: Can receive a str representing a sql statement,
+        a list of str (sql statements), or reference to a template file.
+        Template reference are recognized by str ending in '.sql'
+    :param snowflake_conn_id: Reference to
+        :ref:`Snowflake connection id<howto/connection:snowflake>`
+    :type snowflake_conn_id: str
+    :param autocommit: if True, each command is automatically committed.
```
Comment (JavierLopezT) [inline airflow/providers/snowflake/operators/snowflake.py:171]: Given that the the sqls are meant to be SELECT statements, does it make sense to have this param?

File: airflow/providers/snowflake/operators/snowflake.py [hunk 6]
```diff
@@ -125,3 +126,285 @@ def execute(self, context: Any) -> None:
...
+        self.role = role
+        self.schema = schema
+        self.authenticator = authenticator
+        self.session_parameters = session_parameters
+        self.query_ids = []
+
+
+class SnowflakeValueCheckOperator(_SnowflakeDbHookMixin, SQLValueCheckOperator):
+    """
+    Performs a simple value check using sql code.
```
Comment (JavierLopezT) [inline airflow/providers/snowflake/operators/snowflake.py]: I know that this is copied from https://github.com/apache/airflow/blob/main/airflow/operators/sql.py#L149, but I think it's unclear what the operator really is. I think adding something like 'it checks against a value that you define' would be really useful

File: airflow/providers/snowflake/operators/snowflake.py [hunk 7]
```diff
@@ -125,3 +126,285 @@ def execute(self, context: Any) -> None:
...
+    :type authenticator: str
+    :param session_parameters: You can set session-level parameters at
+        the time you connect to Snowflake
+    :type session_parameters: dict
+    """
+
+    def __init__(
+        self,
+        *,
+        sql: Any,
```
Comment (JavierLopezT) [inline airflow/providers/snowflake/operators/snowflake.py]: Is it really Any? Or union str and list?

File: airflow/providers/snowflake/operators/snowflake.py [hunk 8]
```diff
@@ -125,3 +126,285 @@ def execute(self, context: Any) -> None:
...
+    a certain tolerance of the ones from days_back before.
+
+    This method constructs a query like so ::
+
+        SELECT {metrics_threshold_dict_key} FROM {table}
+        WHERE {date_filter_column}=<date>
+
+    :param table: the table name
+    :type table: str
+    :param days_back: number of days between ds and the ds we want to check
```
Comment (JavierLopezT) [inline airflow/providers/snowflake/operators/snowflake.py:332]: If this is here, I guess `tolerance` and `pass_value` should be in the docstring of `SnowflakeValueCheckOperator`. It seems that for Sphinx is not necessary https://stackoverflow.com/a/2025599/9621172 but certainly it really helps to have them here if you are just reading the code

File: airflow/providers/snowflake/operators/snowflake.py (hunk 4 above)
Comment (potiuk) [reply airflow/providers/snowflake/operators/snowflake.py]: Good point

File: airflow/providers/snowflake/operators/snowflake.py (hunk 5 above)
Comment (potiuk) [reply airflow/providers/snowflake/operators/snowflake.py:171]: The can be also UPDATE statements

File: airflow/providers/snowflake/operators/snowflake.py (hunk 7 above)
Comment (potiuk) [reply airflow/providers/snowflake/operators/snowflake.py]: Good point too.

File: airflow/providers/snowflake/operators/snowflake.py (hunk 4 above)
Comment (denimalpaca) [reply airflow/providers/snowflake/operators/snowflake.py]: Trying to be more DRY. Originally I tried making the `Check` classes inherit from just the `SnowflakeOperator`, but it resulted in an error in the constructors.

File: airflow/providers/snowflake/operators/snowflake.py (hunk 8 above)
Comment (denimalpaca) [reply airflow/providers/snowflake/operators/snowflake.py:332]: Agreed. Those docstrings aren't even in the SQL Operators (and should be).

File: airflow/providers/snowflake/operators/snowflake.py (hunk 7 above)
Comment (denimalpaca) [reply airflow/providers/snowflake/operators/snowflake.py]: Actually just `str` according to the [SQL Operator](https://airflow.apache.org/docs/apache-airflow/stable/_modules/airflow/operators/sql.html#SQLValueCheckOperator) it inherits from.

**Comment by uranusjr** [comment]
This needs to be rebased to latest main.

File: airflow/providers/snowflake/operators/snowflake.py (hunk 4 above)
Comment (JavierLopezT) [reply airflow/providers/snowflake/operators/snowflake.py]: DRY philosophy is great. However, my point was that you don't need a class at all and you can get rid of the class extra layer of abstraction. My comment is inspired by https://www.youtube.com/watch?v=o9pEzgHorH0

Using the same class too in the SnowflakeOperator has been quite an improvement, but I think you can still have just a function in the file, and you can just call it from the operators. Anyway, it's not a big deal and the PR is already approved, so up to u.

Comment (potiuk) [reply airflow/providers/snowflake/operators/snowflake.py]: I wached the presentation :). Entertaining and I mostly agree with it  (not everything :) 

But yeah - in  this case using mixin is definitely over-the-top.

Comment (potiuk) [reply airflow/providers/snowflake/operators/snowflake.py]: Would you mind changing that @denimalpaca ?

Comment (denimalpaca) [reply airflow/providers/snowflake/operators/snowflake.py]: Will have this fixed Tuesday, looks like with the way BaseSQLOperator works, even the function without a class might not be necessary.

File: airflow/providers/snowflake/operators/snowflake.py [hunk 9]
```diff
@@ -100,28 +119,293 @@ def __init__(
...
-            snowflake_conn_id=self.snowflake_conn_id,
-            warehouse=self.warehouse,
-            database=self.database,
-            role=self.role,
-            schema=self.schema,
-            authenticator=self.authenticator,
-            session_parameters=self.session_parameters,
-        )
+    def get_db_hook(self) -> SnowflakeHook:
+        return get_db_hook(self)
```
Comment (kaxil) [inline airflow/providers/snowflake/operators/snowflake.py:123]: ???

**Review by kaxil — APPROVED** [review]

**Pull request merged**